libvast/src/directory.cpp

```diff
@@ -128,32 +128,32 @@ caf::expected<size_t> recursive_size(const std::filesystem::path& root_dir) {
   return total_size;
 }
 
-std::vector<path>
-filter_dir(const path& dir, std::function<bool(const path&)> filter,
+caf::expected<std::vector<std::filesystem::path>>
+filter_dir(const std::filesystem::path& root_dir,
+           std::function<bool(const std::filesystem::path&)> filter,
            size_t max_recursion) {
-  std::vector<path> result;
-  if (max_recursion == 0)
-    return result;
-  for (auto& f : directory(dir)) {
-    switch (f.kind()) {
-      default: {
-        if (!filter || filter(f))
-          result.push_back(f);
-        break;
-      }
-      case path::directory: {
-        // Recurse directories depth-first.
-        auto paths = filter_dir(f, filter, --max_recursion);
-        if (!paths.empty()) {
-          auto begin = std::make_move_iterator(paths.begin());
-          auto end = std::make_move_iterator(paths.end());
-          result.insert(result.end(), begin, end);
-          std::sort(result.begin(), result.end());
-        }
-        break;
-      }
-    }
+  std::vector<std::filesystem::path> result;
+  std::error_code ec{};
+  auto dir = std::filesystem::recursive_directory_iterator(root_dir, ec);
+
+  if (ec)
+    return caf::make_error(ec::filesystem_error, ec.message());
+
+  auto begin = std::filesystem::begin(dir);
+  auto end = std::filesystem::end(dir);
+  result.reserve(std::distance(begin, end));
+  while (begin != end) {
+    const auto current_path = begin->path();
+    const auto current_depth = static_cast<size_t>(begin.depth());
+    if (current_depth >= max_recursion)
+      break;
+
+    if (!filter || filter(current_path))
+      result.push_back(current_path);
+    ++begin;
   }
+
+  result.shrink_to_fit();
```

Comment:
Since this function is not on the critical path, this is useful. Otherwise I'd say let's not stress the allocator.

Reply:
Good observation. I considered it useful in the case where we over allocate up front and a lot of files don't satisfy the filter, so we can shrink our memory footprint.